src/malli/core.cljc

```diff
@@ -1926,12 +1926,151 @@
+                         (or (nil? x) ;; third recursion
+                             (vector? x)
+                             (= 2 (count x))
+                             (int? (nth x 0))
+                             ;; lazy compilation
+                             ((m/validator ConsCell) (nth x 1)))))))))
+
+;; # Goals
+;;
+;; This PR changes how validators are compiled by detecting recursion using
+;; -identify-ref-schema. This function helps us reliably detect ref cycles even in
+;; the presence of dynamic scope (malli's scoping approach for refs) by only
+;; recognizing a ref cycle if the same name _and scope_ are seen.
+;;
+;; This puts us on equal footing with Plumatic Schema's validator compilation.
+;; It is trivial for Schema to detect cycles since they use _global_ scope for
+;; schemas. If they see the same schema again, they can simply reuse the same
+;; validator they were already compiling to "tie the knot" into a recursive validation.
+;;
+;; The goal and purpose of this PR is for ConsCell to compile its validator up-front and once-and-for-all:
+
+(let [rec (volatile! nil)
+      f (fn [x] (or (nil? x)
+                    (and (vector? x)
+                         (= 2 (count x))
+                         (int? (nth x 0))
+                         (@rec (nth x 1)))))]
+  (vreset! rec f)
+  f)
+
+;; This validator does no compilation at runtime and has constant memory usage.
+;; 
+;; # Implementation
+;; 
+;; There are two kinds of refs in malli: lazy and "eager" (or just normal). Lazy ref schemas 
+;; realize their child schema lazily, and eager refs know them upfront. The essential difference
+;; is apparent in their initial validators (see `rf` in the :ref impl):
+
+;; lazy refs only realize their child when the validator is called
+(-memoize (fn [] (schema (mr/-schema (-registry options) ref) options)))
+;;                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+;; eager refs realize their child when compiling their validators
+(when-let [s (mr/-schema (-registry options) ref)] (-memoize (fn [] (schema s options))))
+;;           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+;;
+;; Because of this essential difference, validators for eager refs can be fully compiled, but
+;; lazy refs must  <START FROM HERE, WE CAN FIX THIS>
+;;
+;; ## -validator doesn't take opts
+;; 
+;; A key 
+;;
+;; ## Eager refs
+;; 
+;; Implementation:
+
+(-validator [this]
+  (let [ref-validators *ref-validators*
+        id (-identify-ref-schema this)]
+    (if-some [vol (ref-validators id)]
+      #(@vol %)
+      (let [vol (volatile! nil)
+            s (or (when-let [s (mr/-schema (-registry options) ref)]
+                    (schema s options))
+                  (when-not allow-invalid-refs
+                    (-fail! ::invalid-ref {:type :ref, :ref ref})))
+            f (binding [*ref-validators* (assoc ref-validators id vol)]
+                (-validator s))]
+        (vreset! vol f)
+        f))))
```

Comment:
This impl looks good, thanks! I wonder if the `s` logic could be replaced with something like `deref`?